### ISSUE TITLE
Phase 8: Pilot Testing, ML Integration & Scale Strategy

### ISSUE THREAD

**Issue by holden** [razrfly/volfefe#152]
## Overview

Strategic roadmap for transitioning from "ready to test" to production-grade insider detection. This issue synthesizes the system's current capabilities, identifies gaps, and proposes a phased approach with ML integration.

**Current State**: System Grade B+ (75%)
**Target State**: Production-Grade A+ with ML Enhancement
**Timeline**: 8-12 weeks phased rollout

---

### What's Working ✅
- Core Trinity scoring pipeline (size + timing + wallet age Z-scores)
- Dual-source data ingestion (Subgraph primary, API fallback)
- Flexible JSON-based pattern matching (~20 patterns)
- Investigation candidate queue (~5K candidates)
- Reference case discovery (date-range scanning)
- Admin dashboard (basic functionality)
- Confirmed insider feedback loop
- Real-time TradeMonitor GenServer

### Current Data Scale
| Metric | Current | Target |
|--------|---------|--------|
| Markets synced | ~500 | 15,000+ |
| Trades ingested | ~1M | 30M+ |
| Scored trades | ~500K | 30M+ |
| Confirmed insiders | 50 | 150+ |
| Pattern library | ~20 | 50+ |
| Category coverage | politics-heavy | balanced |

### Performance Baseline
- Precision: ~0.93 (low false alarm rate)
- Recall: ~0.82 (catches most known insiders)
- F1 Score: ~0.87
- Lift: ~12.4x vs random baseline

---

### Polymarket Data Landscape
```
Active markets:     ~500
Historical markets: ~15,000+
Trades per market:  100 - 500,000 (varies by volume)
Total trades:       50-100M+ (full history)
```

### Data Ingestion Strategy

| Tier | Scope | Trades | Time | Use Case |
|------|-------|--------|------|----------|
| Targeted | Top 500 high-volume resolved | ~5M | ~8 hours | Pilot validation |
| Rolling | Last 180 days all categories | ~10M | ~17 hours | Active detection |
| Comprehensive | All resolved markets (2 years) | ~30M | ~50 hours | Robust baselines |
| Full | Complete history | ~100M | ~7 days | ML training |

**Recommendation**: Start with Tier 1 (Targeted), expand to Tier 3 (Comprehensive) for production.

### Ingestion Rate Estimates
- Current rate: ~10,000 trades/minute via Subgraph
- With parallel fetching (4 workers): ~40,000 trades/minute
- Bottleneck: Sequential API calls → need parallel batching

---

### Current Distribution (Imbalanced)

| Category | Polymarket Volume | Known Insiders | Baseline Quality | Priority |
|----------|-------------------|----------------|------------------|----------|
| politics | 80% | 23 (46%) | Good | LOW |
| crypto | 40% | 8 (16%) | Fair | MEDIUM |
| sports | 30% | 3 (6%) | Poor | HIGH |
| entertainment | 20% | 2 (4%) | Poor | HIGH |
| corporate | 25% | 5 (10%) | Fair | MEDIUM |
| science | 15% | 4 (8%) | Poor | HIGH |
| legal | 10% | 3 (6%) | Poor | HIGH |
| other | 5% | 2 (4%) | Poor | HIGH |

### Gap Analysis
1. **Training data bias**: 46% of confirmed insiders from politics → baselines skewed
2. **Missing sub-categories**: Earnings, FDA approvals, product launches, regulatory
3. **No cross-category detection**: Insiders operating across categories undetected
4. **Event taxonomy missing**: Within categories, no sub-type classification

---

### Priority 1: FinBERT for Reference Case Discovery (HIGH IMPACT)
**Problem**: Manual linking of news events to markets
**Solution**: Use existing FinBERT/NER infrastructure to auto-match
```
News: "Trump announces new tariff policy"
       ↓ FinBERT + NER extraction
Entity: Trump, tariff, policy
       ↓ Semantic similarity
Market: "Will Trump implement tariffs?" (0.92 match)
```
**Status**: Infrastructure exists in MultiModelClient, needs wiring to Polymarket module

### Priority 2: Anomaly Detection Model (HIGH IMPACT)
**Problem**: Static thresholds (>2σ = suspicious) miss non-linear patterns
**Solution**: Unsupervised anomaly detection
- **Algorithm**: Isolation Forest or Autoencoder
- **Advantage**: Works with extreme class imbalance (50 positives : 1M negatives)
- **Training**: Learns normal trade distribution, flags outliers
- **Features**: Current 7 Z-scores + 10-15 new features

### Priority 3: Wallet Network Analysis (MEDIUM IMPACT)
**Problem**: Can't detect coordinated insider rings
**Solution**: Graph-based clustering
- Build wallet funding graph from on-chain data
- Cluster by behavior similarity (timing, sizing, win rate)
- Detect linked wallets betting same direction
- Add `ring_id` to investigation candidates

### Priority 4: Trade Sequence Classification (MEDIUM IMPACT)
**Problem**: Point-in-time scoring ignores temporal patterns
**Solution**: LSTM/Transformer on trade sequences
- Input: Last N trades by wallet
- Output: Probability of insider behavior
- Detects: Gradual position building, wash trading

### Priority 5: Ensemble Scoring (LOWER PRIORITY)
**Problem**: Rule-based and ML scores disagree
**Solution**: Calibrated ensemble combining both
- Wait until 150+ confirmed insiders
- Train stacking classifier
- Target: 15% precision improvement

### Feature Engineering Expansion
Current (7 features):
```
size_zscore, timing_zscore, wallet_age_zscore, 
wallet_activity_zscore, price_extremity_zscore,
position_concentration_zscore, funding_proximity_zscore
```

Proposed additions (15+):
```
# Temporal Features
trade_velocity_24h, trade_velocity_7d, position_buildup_pattern

# Wallet Network Features
funding_source_age, degree_centrality, cluster_id, shared_funding_wallets

# Market Context Features
market_volume_rank, liquidity_zscore, spread_at_trade, book_depth_ratio

# Behavioral Features
avg_holding_period, win_streak, loss_recovery_pattern, timing_consistency
```

---

### Data Bottlenecks
| Bottleneck | Current | Impact | Solution |
|------------|---------|--------|----------|
| Ingestion speed | 10K/min | 50h for 30M | Parallel fetching (4x) |
| Scoring pipeline | 1K/sec | 8h for 30M | GenStage parallel processing |
| Baseline recalc | Full table scan | Hours | Incremental updates |
| Storage | 2GB for 1M | 200GB for 100M | PostgreSQL with partitioning |

### Analysis Bottlenecks
| Bottleneck | Current | Impact | Solution |
|------------|---------|--------|----------|
| Pattern validation | Sequential | Hours | Parallel per pattern |
| Candidate review | 15 min each | Days for 5K | Bulk actions + ML triage |
| Investigation | Manual | Slow | Automated evidence gathering |

### UI/UX Bottlenecks
| Bottleneck | Current | Impact | Solution |
|------------|---------|--------|----------|
| Mix task coverage | 22% | Missing workflows | Complete UI integration |
| Candidate queue | No bulk | Slow triage | Select-all, bulk dismiss |
| Evidence gathering | Manual | Time waste | Auto wallet snapshot |

---

### Phase 1: Pilot Validation (Weeks 1-3)
**Goal**: Validate system, grow training data

- [ ] Backtest all 50 confirmed insiders - measure detection rate
- [ ] Analyze false negatives - document why we missed any
- [ ] Run discovery on top 100 high-volume resolved markets
- [ ] Manual review of top 50 candidates - confirm/dismiss each
- [ ] Document new patterns discovered from pilot
- [ ] Expand pattern library based on findings

**Success Criteria**:
- Detection rate >80% on known insiders
- 75+ confirmed insiders (from 50)
- 3-5 new patterns documented

### Phase 2: Scale Data Ingestion (Weeks 4-5)
**Goal**: Comprehensive historical data

- [ ] Implement parallel subgraph fetching (4 concurrent workers)
- [ ] Ingest all resolved markets (last 2 years)
- [ ] Build category-specific baselines with sufficient samples
- [ ] Implement incremental baseline update system
- [ ] Add partitioning for trade_scores table

**Success Criteria**:
- 10M+ trades ingested
- 500+ resolved markets with full data
- Each category has 1000+ trades for baselines

### Phase 3: ML Enhancement (Weeks 6-9)
**Goal**: Augment rule-based with ML scoring

- [ ] Wire FinBERT to reference case discovery
- [ ] Build Isolation Forest anomaly model
- [ ] Expand feature set (7 → 22 features)
- [ ] Train classifier on expanded confirmed insiders (when 150+)
- [ ] Implement ensemble scoring (rules + ML)
- [ ] Add `ml_anomaly_score` and `ml_confidence` fields

**Success Criteria**:
- 15% improvement in precision
- 10% improvement in recall
- F1 score >0.90

### Phase 4: Network Analysis (Weeks 10-12)
**Goal**: Detect coordinated insider rings

- [ ] Build wallet funding graph from on-chain data
- [ ] Implement wallet clustering by behavior similarity
- [ ] Detect linked wallets betting same direction
- [ ] Add `ring_id` and `cluster_confidence` to candidates
- [ ] Cross-market investigation view

**Success Criteria**:
- 5+ wallet clusters identified
- 2+ confirmed insider rings discovered

### Phase 5: UI/UX Production Polish (Weeks 10-12, parallel)
**Goal**: 100% mix task coverage, efficient triage

- [ ] Complete remaining 21 mix task UI integrations
- [ ] Bulk candidate actions (select 10, dismiss all)
- [ ] ML-assisted queue prioritization
- [ ] Automated evidence snapshot per candidate
- [ ] Export investigation reports

**Success Criteria**:
- 27/27 mix tasks in UI (100%)
- <2 min average investigation time (from 15 min)
- <5% false positive rate

---

### Current (Sufficient for Phase 1-2)
- PostgreSQL (scales to 30M trades)
- Single Elixir node
- Subgraph API access

### Phase 3+ Additions
- Redis for caching baselines and session state
- GenStage/Broadway for parallel processing
- Cloud GPU for model training (1-2 days/model)
- Vector DB for semantic search (if adding embedding-based matching)

### Storage Projections
| Scale | Trades | PostgreSQL Size | Notes |
|-------|--------|-----------------|-------|
| Current | 1M | ~2GB | Comfortable |
| Pilot | 5M | ~10GB | Fine |
| Production | 30M | ~60GB | Standard SSD |
| Full History | 100M | ~200GB | May need archival strategy |

---

### Detection Quality
| Metric | Current | Phase 1 Target | Production Target |
|--------|---------|----------------|-------------------|
| Precision | 0.93 | 0.95 | 0.97 |
| Recall | 0.82 | 0.85 | 0.90 |
| F1 Score | 0.87 | 0.90 | 0.93 |
| False Positive Rate | ~7% | <5% | <3% |

### Operational Efficiency
| Metric | Current | Target |
|--------|---------|--------|
| Investigation time | 15 min | <2 min |
| Candidate queue size | 5K | Manageable (<500 high priority) |
| Ingestion speed | 10K/min | 40K/min |
| UI task coverage | 22% | 100% |

### Training Data
| Metric | Current | Target |
|--------|---------|--------|
| Confirmed insiders | 50 | 150+ |
| Category balance | 46% politics | <30% any single |
| Pattern library | 20 | 50+ |

---

### Technical Risks
| Risk | Probability | Impact | Mitigation |
|------|-------------|--------|------------|
| Class imbalance breaks ML | Medium | High | Use anomaly detection, not classification |
| Subgraph rate limits | Low | Medium | Exponential backoff, caching |
| PostgreSQL performance at scale | Low | High | Add indexes, partition tables |
| Feature engineering complexity | Medium | Medium | Start with current 7, expand incrementally |

### Business Risks
| Risk | Probability | Impact | Mitigation |
|------|-------------|--------|------------|
| Insiders adapt to detection | Medium | High | Ensemble approach, regular pattern updates |
| False positives damage credibility | Medium | High | Conservative thresholds, ML calibration |
| Training data stays small | Medium | Medium | Incentivize discovery, expand reference cases |

---

### When to Add ML?
- **Now (FinBERT)**: Already have infrastructure, low effort
- **After 75 insiders (Anomaly)**: Minimum viable training set
- **After 150 insiders (Classifier)**: Robust supervised learning
- **After network data (Graph ML)**: When funding graph built

### When to Scale Data?
- **After pilot validation**: Confirm patterns work on small set first
- **Before ML training**: Need 10M+ trades for robust feature distributions

### When to Invest in Infrastructure?
- **Redis/GenStage**: When processing >10M trades regularly
- **Vector DB**: If adding embedding-based semantic search
- **Dedicated GPU**: Only for model training, not inference

---

## 11. Next Steps

**Immediate (This Week)**:
1. [ ] Run `mix polymarket.backtest` on all 50 confirmed insiders
2. [ ] Document detection rate and false negatives
3. [ ] Identify top 100 resolved high-volume markets for pilot

**Short-term (Weeks 1-2)**:
1. [ ] Ingest pilot market data (~5M trades)
2. [ ] Run discovery, generate candidate queue
3. [ ] Manual review of top 50 candidates

**Medium-term (Weeks 3-5)**:
1. [ ] Implement parallel ingestion
2. [ ] Wire FinBERT to reference case matching
3. [ ] Expand to comprehensive data set

---

## Related Issues
- Builds on #134 (Phase 7 Dashboard)
- Implements suggestions from #133 (Trinity Weighting)
- Extends #131 (Production Hardening)
- Addresses gaps in #151 (Pilot Testing)
- Connects to #122 (ML Backtesting)

## Labels
`enhancement` `ml` `strategy` `phase-8`
